scripts/generate_refcats_gen3.py

```diff
@@ -0,0 +1,248 @@
+        if id > last_id + 1:
+            # last_id ends the range, and id starts a new one
+            ranges.append(_make_range(range_start, last_id))
+            range_start = id
+    ranges.append(_make_range(range_start, id))  # id is last element in asc_ids
+    return ranges
+
+
+id_ranges = _to_ranges(shards)
+
+
+########################################
+# Transfer shards
+
+DEST_DIR = "${AP_VERIFY_HITS2015_DIR}/preloaded/"
+STD_REFCAT = "refcats"
+DEST_RUN = "refcats/imported"
+
+
+def _rename_dataset_type(type, name):
+    """Create a DatasetType that differs from an existing one in name only.
+
+    Parameters
+    ---------
+    type : `lsst.daf.butler.DatasetType`
+        The type to rename.
+    name : `str`
+        The new name to adopt.
+
+    Returns
+    -------
+    new_type : `lsst.daf.butler.DatasetType`
+        The new DatasetType.
+    """
+    return DatasetType(name, type.dimensions, type.storageClass, type.parentStorageClass)
+
+
+src_repo = Butler(args.src_dir, collections=args.src_collection, writeable=False)
+dest_repo = Butler(DEST_DIR, run=DEST_RUN, writeable=True)
+
+
+def _remove_refcat_run(butler, run):
+    """Remove a refcat run and any references from a repository.
+
+    Parameters
+    ----------
+    butler : `lsst.daf.butler.Butler`
+        The repository from which to remove ``run``.
+    run : `str`
+        The run to remove, if it exists.
+    """
+    try:
+        refcat_runs = butler.registry.getCollectionChain(STD_REFCAT)
+        if DEST_RUN in refcat_runs:
+            new_runs = list(refcat_runs)
+            new_runs.remove(DEST_RUN)
+            butler.registry.setCollectionChain(STD_REFCAT, new_runs)
+    except (lsst.daf.butler.MissingCollectionError, TypeError):
+        pass  # No STD_REFCAT chain; nothing to do
+
+    try:
+        butler.removeRuns([DEST_RUN], unstore=True)
+    except lsst.daf.butler.MissingCollectionError:
+        pass  # Already removed; nothing to do
+
+
+logging.info("Preparing destination repository %s...", DEST_DIR)
+_remove_refcat_run(dest_repo, DEST_RUN)
+dest_repo.registry.registerCollection(DEST_RUN, CollectionType.RUN)
+for src_cat, dest_cat in REFCATS.items():
+    src_type = src_repo.registry.getDatasetType(src_cat)
+    dest_type = _rename_dataset_type(src_type, dest_cat)
+    dest_repo.registry.registerDatasetType(dest_type)
+dest_repo.registry.refresh()
+
+logging.info("Searching for refcats in %s...", args.src_dir)
+query = f"htm{HTM_LEVEL} in ({','.join(id_ranges)})"
+datasets = []
+for src_ref in src_repo.registry.queryDatasets(REFCATS.keys(), where=query, findFirst=True):
+    src_type = src_ref.datasetType
+    dest_type = _rename_dataset_type(src_type, REFCATS[src_type.name])
+    dest_ref = DatasetRef(dest_type, src_ref.dataId)
+    datasets.append(FileDataset(path=src_repo.getURI(src_ref), refs=dest_ref))
+
+logging.info("Copying refcats...")
+dest_repo.ingest(*datasets, transfer="copy")
```

Comment:
Why specifically use `copy` here? Is duplicating the files necessary?

Reply:
Absolutely. The `ap_verify` dataset must be completely self-contained, and not using links to a repo that might be on a different machine.

Reply:
Please put a comment above this line to that end.